fix: Recreate group in case redis flush

pbonneaudiabolocom opened the pull request:
# Description

When using redis with group, if a flush is done by another parties or the stream is destroyed, faststream enter in an infinite loop, trying to get the next element but not finding the expected group.

Fixes #2611

## Type of change

Please delete options that are not relevant.

- [ ] Documentation (typos, code examples, or any documentation updates)
- [X] Bug fix (a non-breaking change that resolves an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a fix or feature that would disrupt existing functionality)
- [ ] This change requires a documentation update

## Checklist

- [X] My code adheres to the style guidelines of this project (`just lint` shows no errors)
- [X] I have conducted a self-review of my own code
- [X] I have made the necessary changes to the documentation
- [X] My changes do not generate any new warnings
- [x] I have added tests to validate the effectiveness of my fix or the functionality of my new feature
- [X] Both new and existing unit tests pass successfully on my local environment by running `just test-coverage`
- [x] I have ensured that static analysis tests are passing by running `just static-analysis`
- [ ] I have included code examples to illustrate the modifications

Lancetnik commented on faststream/redis/subscriber/usecases/stream_subscriber.py:
```diff
@@ -129,14 +132,42 @@ def read(
+                        )
+                    except ResponseError as e:
+                        if "NOGROUP" in str(e):
+                            await self._create_group()
```
Should we change current `last_id` read position at stream recreation?

pbonneaudiabolocom committed:
fix: Recreate group in case redis flush
 - adding a test
 - reset properly the id
 - adding another rare case when group is tempered by another system, and IDs gets desynchronized. (AI recommendation)
 - fixing group creation in autoclaim + get_one message and iter.

JonathanSerafini commented on faststream/redis/subscriber/usecases/stream_subscriber.py:
```diff
@@ -96,38 +96,145 @@ async def start(self) -> None:
+                    ...,
+                ],
+            ]:
+                async def read_from_group() -> tuple[
```
i'm thinking at this point you may as well just make `def read` an async function and do without the nested `read_from_group` it may make things more readable ?

JonathanSerafini replied:
this type annotation is likely the first conflict, you can use `ReadResponse` as a replacement

pbonneaudiabolocom commented:
@JonathanSerafini Would you think it' possible to merge it ?

Last time it was working fine, but the more the system progress, the more I have to redo the job...

pbonneaudiabolocom committed:
fix: Recreate group in case redis flush
 - adding a test
 - reset properly the id
 - adding another rare case when group is tempered by another system, and IDs gets desynchronized. (AI recommendation)
 - fixing group creation in autoclaim + get_one message and iter.

Lancetnik commented:
Hi @pbonneaudiabolocom

Thanks for the persistence on this and for the detailed reasoning — both the problem and the trade-offs you've laid out are real, and I appreciate you engaging with the discussion rather than just pushing the patch through.

I want to expand on @powersemmi's point, because I don't think we framed the "why" clearly enough.
FastStream's stance on infrastructure management is intentional and consistent across brokers: the application is a consumer of infrastructure, not its owner. The startup-time queue/group/stream creation we do today is explicitly a developer-experience convenience for local environments, and we recommend disabling it in production. The analogy we usually reach for is databases — an application doesn't create or repair its own tables at runtime; that responsibility belongs to migrations, IaC, or the platform team. Streams and consumer groups sit in the same category. Silently recreating a group that disappeared means the application is now papering over an incident (a FLUSHALL, a misconfigured retention, an operator mistake) that someone almost certainly needs to know about.
So the correct behavior on NOGROUP really is to surface the error loudly, not to self-heal. "Crash and get restarted by the orchestrator with alerts firing" is a feature in this model, not a regression.
That said, your follow-up questions are the right ones, and I don't want to leave you without a path forward:

> Can we capture/handle exceptions raised by FastStream somewhere?
> Can we call the group creation / subscriber restart?

Both yes, and I think this is where the PR could actually land in a way that fits the framework:

Make the failure observable and typed. Today the loop swallows NOGROUP and spins. The minimal correct fix is to let that ResponseError (or a wrapped, more specific exception like StreamGroupNotFoundError) propagate out of the subscriber instead of being retried indefinitely. That alone solves the "infinite loop until restart" symptom from #2611 — the process fails fast, the orchestrator restarts it, the startup-time group creation re-runs, and you're back online. No multi-publisher app dies silently.

Give users a hook, not automatic recovery. If a user explicitly wants the self-healing behavior for their environment, exposing it as opt-in (an error handler, a lifespan callback, or a documented pattern using FastStream's existing exception middleware) is much more in line with the framework's philosophy than baking it into the default read path. The user owns the policy; FastStream just gives them the seam.

If you'd be open to reshaping the PR along those lines — propagating a clear, catchable exception instead of recreating the group inline — I think there's a much shorter path to merge here, and it would actually close #2611 more robustly than the current approach (which only handles the Redis-flush case, not e.g. the stream being deleted while groups exist elsewhere).

Happy to discuss the exception type and where exactly it should be raised if that direction works for you.

powersemmi commented:
I fix this here https://github.com/ag2ai/faststream/pull/2855

pbonneaudiabolocom commented:
thank you both for your time.

Our team suggest to add a warning in case someone trust Faststream to create the group, reminding him that's it's not production ready / advised)

I'll follow up in the other PR :)
